quinn-udp/src/windows.rs

```diff
@@ -111,22 +300,125 @@ impl UdpSocketState {
         bufs: &mut [IoSliceMut<'_>],
         meta: &mut [RecvMeta],
     ) -> io::Result<usize> {
-        // Safety: both `IoSliceMut` and `MaybeUninitSlice` promise to have the
-        // same layout, that of `iovec`/`WSABUF`. Furthermore `recv_vectored`
-        // promises to not write unitialised bytes to the `bufs` and pass it
-        // directly to the `recvmsg` system call, so this is safe.
-        let bufs = unsafe {
-            &mut *(bufs as *mut [IoSliceMut<'_>] as *mut [socket2::MaybeUninitSlice<'_>])
-        };
-        let (len, _flags, addr) = socket.0.recv_from_vectored(bufs)?;
-        meta[0] = RecvMeta {
-            len,
-            stride: len,
-            addr: addr.as_socket().unwrap(),
-            ecn: None,
-            dst_ip: None,
-        };
-        Ok(1)
+        if let Some(wsa_recvmsg_ptr) = self.wsa_recvmsg_ptr {
+            // we cannot use [`socket2::MsgHdrMut`] as we do not have access to inner field which holds the WSAMSG
+            let mut ctrl_buf = cmsg::Aligned([0; CMSG_LEN]);
+            let mut source: WinSock::SOCKADDR_INET = unsafe { mem::zeroed() };
+            let mut data = WinSock::WSABUF {
+                buf: bufs[0].as_mut_ptr(),
+                len: bufs[0].len() as _,
+            };
+
+            let ctrl = WinSock::WSABUF {
+                buf: ctrl_buf.0.as_mut_ptr(),
+                len: ctrl_buf.0.len() as _,
+            };
+
+            let mut wsa_msg = WinSock::WSAMSG {
+                name: &mut source as *mut _ as *mut _,
+                namelen: mem::size_of_val(&source) as _,
+                lpBuffers: &mut data,
+                Control: ctrl,
+                dwBufferCount: 1,
+                dwFlags: 0,
+            };
+
+            // FIXME add Safety: ?
+            let mut len = 0;
+            unsafe {
+                let rc = (wsa_recvmsg_ptr)(
+                    socket.0.as_raw_socket() as usize,
+                    &mut wsa_msg,
+                    &mut len,
+                    ptr::null_mut(),
+                    None,
+                );
+                if rc == -1 {
+                    return Err(io::Error::last_os_error());
+                }
+            }
+
+            // FIXME add Safety: ?
+            let addr = unsafe {
+                let (_, addr) = socket2::SockAddr::try_init(|addr_storage, len| {
```

Comment:
Would `SockAddr::new` be a bit more concise here and below? I think `try_init` is intended for cases when you already have a (typically foreign) function that takes an output pointer.

Reply:
i will take a look and see if it works